java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/UIDefaultsIconMetadata.java

```diff
@@ -0,0 +1,359 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.nbcode.integration;
+
+import java.awt.Component;
+import java.awt.EventQueue;
+import java.awt.Graphics;
+import java.awt.HeadlessException;
+import java.awt.Image;
+import java.awt.MediaTracker;
+import java.awt.Toolkit;
+import java.awt.Transparency;
+import java.awt.image.BufferedImage;
+import java.awt.image.ColorModel;
+import java.awt.image.ImageObserver;
+import java.awt.image.WritableRaster;
+import java.beans.PropertyChangeEvent;
+import java.beans.PropertyChangeListener;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.Enumeration;
+import java.util.Hashtable;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.swing.Icon;
+import javax.swing.ImageIcon;
+import javax.swing.UIDefaults;
+import javax.swing.UIDefaults.ActiveValue;
+import javax.swing.UIDefaults.LazyValue;
+import javax.swing.UIManager;
+import org.openide.modules.OnStart;
+import org.openide.util.ImageUtilities;
+
+/**
+ * Goes through UIDefaults and replaces Icons and Images with variants that
+ * return image's URL from the 'url' property.
+ * @author sdedic
+ */
+@OnStart
+public class UIDefaultsIconMetadata implements Runnable {
+    private static final Logger LOG = Logger.getLogger(UIDefaultsIconMetadata.class.getName());
+    
+    // for diagnostic purposes only
+    public static final String PROP_ORIGINAL_ICON = "originalIcon"; // NOI18N
+    // for diagnostic purposes only
+    public static final String PROP_ORIGINAL_IMAGE = "originalImage"; // NOI18N
+    
+    private static final String URN_DEFAULTS_PREFIX = "uidefaults:"; // NOI18N
+    
+    private static Object wrapImage(String key, Image image, Icon icon, Object... meta) {
+        Object o = image.getProperty(ImageUtilities.PROPERTY_ID, null); // NOI18N
+        if (o instanceof URL) {
+            return image;
+        }
+        
+        o = image.getProperty(ImageUtilities.PROPERTY_ID, null);
+        if (o instanceof String || o instanceof URI) {
+            return image;
+        } else {
+            
+            Hashtable props = new Hashtable();
+            for (int i = 0; i < meta.length; i+= 2) {
+                props.put(meta[i], meta[i+1]);
+            }
+            props.put(ImageUtilities.PROPERTY_ID, URN_DEFAULTS_PREFIX + key);
+            return createNew(image, icon, props);
```

Comment:
OK, I'm actually confused about what UIDefaultsIconMetadata does. Is it trying to use the UIManager key to identify the icon?

Reply:
OK, I should change the (now after the changes) obsolete javadoc. The only purpose of that is to 'somehow add metadata' to the Icons and Images available from `UIManager.getIcon/Image()`. It does so by wrapping the originals into a delegate that provides appropriate value(s) from `Image.getProperty` - and that's all.

Reply:
I maybe do not understand the question:  The client receives an Image instance. That instance may be possibly badged from some base Image instance the client has no access to. How could the client compute that the (base) image comes from UIDefault key xxx when it does not know the base image instance, or how the base image has been obtained ?

Reply:
Then there is the Icon > Image conversion that happens sometimes in the infrastructure (i.e. FolderNode) which also needs (somehow) to carry over the ID; in this case, the converting code does not even know that someone will query an ID from the resulting image (or an image badged from that).

Reply:
> The client receives an Image instance. That instance may be possibly badged from some base Image instance the client has no access to. How could the client compute that the (base) image comes from UIDefault key xxx when it does not know the base image instance, or how the base image has been obtained (maybe even not from UIDefaults) ?

I think this is backwards. By the time you have an Image or an Icon, it's already too late to try to substitute another icon. The substitution needs to happen when you _load_ the Image or the Icon. That's when the code loading the icon should be supplying an ID.

> Then there is the Icon > Image conversion that happens sometimes in the infrastructure (i.e. FolderNode) which also needs (somehow) to carry over the ID

There was a similar problem when I implemented HiDPI icon support in ImageUtilities. The solution was to make sure that ImageUtilities.icon2image and ImageUtilities/image2icon were reversible. So the metadata is carried over.

Reply:
OK, so as long as all of this only ever runs in the headless LSP use case, I have a lot fewer concerns. Adding the string id field to ImageUtilities also seems fine, as long as there is no extra instance wrapping in the regular IDE.

> The use case for headless LSP server is different. The icon, for example in our case, resides in a completely other binary/product with a completely different loading and painting mechanism.

So the icons to be replaced are not icons in the NetBeans codebase? Could you give an example of such an icon, perhaps with a screenshot? (A screenshot would make it a lot easier for me to understand what, exactly, the LSP server is...) What are we replacing these icons with?

> The imageID stuff supplies some ID for non-URL images, most notably those from UIDefaults (not loaded by loadImage)

Right, but UIDefaults icons can be replaced by changing the UIDefaults itself, no?

Reply:
> The use case for **headless** LSP server is different. 

My general point is that the use case for the headless LSP server is different to what we have now, but that the IDE would benefit from fixing the process in a similar way.

> For Images obtained as `ImageUtilities.loadImage` it's rather easy as there's the URL (which can be mangled by localization etc, but that can be handled).

And the URL has _been_mangled to address branding, and dark variants, and sizes, and hidpi, and svg conversion ... Some of which interfere badly with each other, and spread files (some duplicates IIRC) all over the code base.  For the purposes of loading an _icon_ at least, there are benefits to not mangling based on URL in the first place.